ssh: Drop challenge_unknown_host_preconnect hack

martinpitt opened the pull request:
Conceptually we always want the bridge to send a challenge for unknown hosts, to keep the protocol non-magical and well-defined.

This does not break existing cockpit/ws containers for existing bastion host setups: This does not change the behaviour of cockpit-bridge, only cockpit-ssh. And cockpit/ws containers always include their matching cockpit-ssh (with possibly the old behaviour), they don't talk to a remote cockpit-ssh directly. But let's make double-sure.

 - [x] Fix cockpit-ws to respond to host key challenges: PR #10442
 - [x] Make double-sure that old (e. g. RHEL 7) cockpit/ws can talk to a remote with this new behaviour.
 - [x] Check the [ManageIQ](https://github.com/ManageIQ/manageiq) code base to make sure this does not break functionality.
 - [x] Check the  [ovirt-cockpit-sso](https://github.com/oVirt/ovirt-cockpit-sso) to make sure this does not break functionality.
 - [x] Bump dependencies to fixed cockpit-ws
 - [x] for RHEL/CentOS 7, either wait until the fix lands there: PR #10585

martinpitt commented:
As described, this cannot land just yet. I just want to run integration tests to see which ones fail now. I mostly expect the `testDirectLogin` one, but let's see.

martinpitt commented:
Locally, these three check-multi-machine tests  fail: `testDirectLogin`, `testUrlRoot`, `testFIPS`. I fully expected the first two (confirming that our tests spot the regression), the latter is just an innocent victim of the same problem. I'll work on the cockpit-ws fix next.

martinpitt commented:
Fixed memory leak in test-sshoption.

martinpitt commented:
@stefwalter : If you have some time, I'd appreciate a review from you as well. Sanne and I are slowly getting the hang of the protocol bits, but not quite up to your level yet. In particular, I wonder if we miss something here. I added a TODO item to check ManageIQ and ovirt-ssh (like in PR #10143).

stefwalter commented on src/ws/cockpitauth.c:
```diff
@@ -780,6 +780,21 @@ on_transport_control (CockpitTransport *transport,
+
+          /* return a negative answer; we handle unknown hosts interactively, or want to fail on them */
+          g_debug ("received x-host-key authorize challenge");
+          send_authorize_reply (session->transport, cookie, "");
```
According to cockpitsshrelay.c we should be setting "* invalid key" if we want it to fail. Is this correct? Either way this is begging for documentation at this point. Can we update doc/authorize.md?

martinpitt replied:
So `""` is the legacy one and `"* invalid key"` the current one ([code](https://github.com/cockpit-project/cockpit/blob/master/src/ssh/cockpitsshrelay.c#L318)). Indeed I noticed that with values `""` and `"*"` (which I also tried), what actually arrives in cockpit-ssh is always `* invalid key`. So I'll change it to that and document this, too.

martinpitt replied:
@stefwalter: Actually, [challenge_for_auth_data()](https://github.com/cockpit-project/cockpit/blob/master/src/ssh/cockpitsshrelay.c#L299) only treats `""` as invalid . With `"* invalid key"`, challenge_for_knownhosts_data() returns with a value and triggers the

    authorize challenge reported key for 127.0.0.1:43183 which is not known to cockpit_is_host_known()

warning. I can silence this, but this would mean more special cases. What's wrong with using an empty response?

stefwalter replied:
Lets use empty then, but:

- Use it consistently
- Document it.

martinpitt replied:
I [documented it](https://github.com/martinpitt/cockpit/blob/drop-challenge-preconnect-hack/doc/authentication.md) now (separate commit), and also changed the unit tests to use an `INVALID_KEY` macro, so that it's both more obvious what it is, and we can also change it centrally if we want to.

martinpitt commented:
I checked ovirt-cockpit-sso. cockpit-auth-ovirt is a wrapper around cockpit-ssh, used as a custom auth command for cockpit-ws. It first authenticates to ovirt using a token (completely unrelated to ssh), and then `execlp()`s cockpit-ssh to deal with the SSH authentication. Thus if cockpit-ws → cockpit-ssh works, then cockpit-ws → cockpit-auth-ovirt → cockpit-ssh should work in exactly the same way.

martinpitt commented:
ManageIQ's [cockpit-auth-miq](https://github.com/ManageIQ/manageiq/blob/master/tools/cockpit/cockpit-auth-miq) is a shim very similar to cockpit-auth-ovirt. It execs cockpit-ssh as the very last thing, and does its own protocol before which isn't affected by this change. It is just used to be [set in cockpit.conf](https://github.com/ManageIQ/manageiq/blob/master/lib/miq_cockpit.rb#L118) as a login command for -ws, so by the same reasoning: if cockpit-ws → cockpit-ssh works, then cockpit-ws → cockpit-auth-miq → cockpit-ssh should work in exactly the same way.

martinpitt commented:
Need to special-case the rhel-7-*-distropkg tests. We don't aim to land this at RHEL/CentOS 7 anyway, this is just "future cleanup". But our tests check the new cockpit-ssh (via cockpit-dashboard) against the old ws.

martinpitt commented:
Pushed again with this interdiff and an explanation in the changelog:

```diff
diff --git a/test/verify/check-multi-machine b/test/verify/check-multi-machine
index a7fc771c8..c94084283 100755
--- a/test/verify/check-multi-machine
+++ b/test/verify/check-multi-machine
@@ -288,6 +288,12 @@ class TestMultiMachine(MachineCase):
         b.wait_visible("#option-group")
         b.wait_not_visible("#server-group")
         b.wait_visible("#login-error-message")
+        if m.image in ['rhel-7-5-distropkg', 'rhel-7-6-distropkg']:
+            # cockpit-ws in 7.5/7.6 misinterprets x-host-key challenge (PR #10430)
+            # stop the test here, the other "unknown host key" checks below fail, too
+            b.wait_in_text("#login-error-message", "Wrong user")
+            return
+
         b.wait_in_text("#login-error-message", "Host is unknown")
 
         m.execute("echo '\n[SSH-Login]\nallowUnknown = true' >> /etc/cockpit/cockpit.conf")
```

@Gundersanne: Can you please re-review?

stefwalter commented:
> Pushed again with this interdiff and an explanation in the changelog:

This means there's an API break, and/or compatibility problem. I have doubts about landing this if the integration tests require modification.

martinpitt commented:
@stefwalter: Right, see the changelog:

    However, this does break when mixing a newer cockpit-ssh with an older
    cockpit-ws, which gets exposed in the two `checkDirectLogin()` tests on
    rhel-7-*-distropkg. This change must not be backported to RHEL/CentOS 7
    without backporting the corresponding "ws: Handle x-host-key challenge"
    fix.

We stop releasing new upstream versions from master to RHEL/CentOS 7 now, so I don't think it's a big deal. (We already disabled other tests on rhel-7-*-distropkg for similar reasons). Soon we'll throw out a lot of stuff from master and only maintain the [stable back branch](https://github.com/cockpit-project/cockpit/tree/rhel-7.6) for RHEL 7. If you object heavily, we can also only land the cockpit-ws fix, and postpone the cleanup until after that "big cleanup", but practically it doesn't make much difference IMHO.

stefwalter commented:
> We stop releasing new upstream versions from master to RHEL/CentOS 7 now, so I don't think it's a big deal. (We already disabled other tests on rhel-7-*-distropkg for similar reasons). Soon we'll throw out a lot of stuff from master and only maintain the stable back branch for RHEL 7. If you object heavily, we can also only land the cockpit-ws fix, and postpone the cleanup until after that "big cleanup", but practically it doesn't make much difference IMHO.

Two questions:

 * Does the regression *only* affect direct communication between ```cockpit-ws``` -> ```cockpit-ssh```
 * If so, can we address this regression with an appropriate change in the spec file and debian rules?

martinpitt commented:
> Does the regression only affect direct communication between cockpit-ws -> cockpit-ssh

Yes. When going via cockpit-bridge, i. e. from all the JS part, the behaviour is unchanged as the [ssh manifest](https://github.com/cockpit-project/cockpit/blob/master/pkg/ssh/manifest.json.in) hardcodes the `COCKPIT_SSH_CHALLENGE_UNKNOWN_HOST_PRECONNECT=true` hack.

> If so, can we address this regression with an appropriate change in the spec file and debian rules?

In debian/rules, no problem. In the spec file we could bump the dependency for Fedora and rhel-x, but not for rhel/centos 7, at least not until the ws fix actually lands somewhere.

Anyway, this is just a nice cleanup to have. The really important prerequisite is the ws fix, which by itself  does not break anything (in fact, it's rather inert until we land this cleanup). I split this out into #10442, so we can land this now. I updated the TODO items above for the dependencies and waiting for either fixing this in RHEL or  dropping.

So at least the test results here prove that #10442 works (as this cannot be tested by itself).

martinpitt commented:
I rebased and added the Conflicts:. These should work everywhere except rhel-7-*-distropkg. PR #10445 backports the ws fix, but we need to release and land it in 7.6.z before this will go green and can land.
